Source/Core/Core/WiiUtils.cpp

```diff
@@ -684,4 +684,89 @@ UpdateResult DoDiscUpdate(UpdateCallback update_callback, const std::string& ima
   DiscIO::NANDContentManager::Access().ClearCache();
   return result;
 }
+
+bool CheckNAND(IOS::HLE::Kernel& ios)
+{
+  bool bad = false;
+  const auto es = ios.GetES();
+
+  // Check for NANDs that were used with old Dolphin versions.
+  if (File::Exists(Common::RootUserPath(Common::FROM_CONFIGURED_ROOT) + "/sys/replace"))
+  {
+    WARN_LOG(CORE, "CheckNAND: NAND was used with old versions, so it is likely to be damaged");
+    bad = true;
+  }
+
+  for (const u64 title_id : es->GetInstalledTitles())
+  {
+    // Check for missing title sub directories.
+    if (!File::IsDirectory(Common::GetTitleContentPath(title_id, Common::FROM_CONFIGURED_ROOT)))
+    {
+      WARN_LOG(CORE, "CheckNAND: Missing content directory for title %016" PRIx64, title_id);
+      bad = true;
+    }
+    if (!File::IsDirectory(Common::GetTitleDataPath(title_id, Common::FROM_CONFIGURED_ROOT)))
+    {
+      WARN_LOG(CORE, "CheckNAND: Missing data directory for title %016" PRIx64, title_id);
+      bad = true;
+    }
+
+    // Check for incomplete title installs (missing ticket, TMD or contents).
+    const auto ticket = DiscIO::FindSignedTicket(title_id);
+    if (!IOS::ES::IsDiscTitle(title_id) && !ticket.IsValid())
+    {
+      WARN_LOG(CORE, "CheckNAND: Missing ticket for title %016" PRIx64, title_id);
+      bad = true;
+    }
+
+    const auto tmd = es->FindInstalledTMD(title_id);
+    if (!tmd.IsValid())
+    {
+      WARN_LOG(CORE, "CheckNAND: Missing TMD for title %016" PRIx64 " (warning)", title_id);
```

Comment:
Should an invalid TMD be a reason for the NAND to be marked as bad? I'm not really certain if this should ever be the case in a consistent NAND...

Also, why do you write a WARN log entry that ends with "(warning)"? :D

Reply:
Hmm, you're right. Thinking more about it, I don't think the TMD should ever be missing or invalid for titles.

I originally made the check ignore missing TMDs because Dragon Quest X update titles have no TMD in their content directory, but I think that may actually be a bug with import finalisation :D

Reply:
Just checked with skyeye, looks like ES *can* leave titles without a TMD if you cancel an import...

Reply:
Suppose its ok then, if thats a valid state.